pkg/cloudprovider/provider/azure/provider.go

```diff
@@ -0,0 +1,615 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve status for VM %q: %s", machine.Spec.Name, err.Error())
+	}
+
+	return &azureVM{vm: &vm, ipAddresses: ipAddresses, status: status}, nil
+}
+
+func (p *provider) Delete(machine *v1alpha1.Machine, instance instance.Instance) error {
+	config, _, err := p.getConfig(machine.Spec.ProviderConfig)
+	if err != nil {
+		return cloudprovidererrors.TerminalError{
+			Reason:  v1alpha1.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	// TODO: Disassociate and remove the resources in reverse order, to prevent
+	// orphaned NICs, IPs and disks in case the operation is interrupted in the middle.
+	glog.Infof("deleting VM %q", machine.Name)
+	if err = deleteVMsByMachineUID(context.TODO(), config, machine.UID); err != nil {
+		return fmt.Errorf("failed to remove public IP addresses of machine %q: %s", machine.Name, err)
+	}
+
+	glog.Infof("deleting disks of VM %q", machine.Name)
+	if err = deleteDisksByMachineUID(context.TODO(), config, machine.UID); err != nil {
+		return fmt.Errorf("failed to remove disks of machine %q: %s", machine.Name, err)
+	}
+
+	glog.Infof("deleting network interfaces of VM %q", machine.Name)
+	if err = deleteInterfacesByMachineUID(context.TODO(), config, machine.UID); err != nil {
+		return fmt.Errorf("failed to remove network interfaces of machine %q: %s", machine.Name, err)
+	}
+
+	glog.Infof("deleting public IP addresses of VM %q", machine.Name)
+	if err = deleteIPAddressesByMachineUID(context.TODO(), config, machine.UID); err != nil {
+		return fmt.Errorf("failed to remove public IP addresses of machine %q: %s", machine.Name, err)
+	}
+
+	return nil
+}
+
+func getVMByUID(ctx context.Context, c *config, uid types.UID) (*compute.VirtualMachine, error) {
+	vmClient, err := getVMClient(c)
+	if err != nil {
+		return nil, err
+	}
+
+	list, err := vmClient.ListAll(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var allServers []compute.VirtualMachine
+
+	for {
+		allServers = append(allServers, list.Values()...)
+		if list.NotDone() {
+			list.Next()
+		} else {
+			break
+		}
+	}
+
+	for _, vm := range allServers {
+		if vm.Tags != nil && *vm.Tags[machineUIDTag] == string(uid) {
+			return &vm, nil
+		}
+	}
+
+	return nil, cloudprovidererrors.ErrInstanceNotFound
+}
+
+func getVMStatus(ctx context.Context, c *config, vmName string) (instance.Status, error) {
+	vmClient, err := getVMClient(c)
+	if err != nil {
+		return instance.StatusUnknown, err
+	}
+
+	iv, err := vmClient.InstanceView(ctx, c.ResourceGroup, vmName)
+	if err != nil {
+		return instance.StatusUnknown, fmt.Errorf("failed to get instance view for machine %q: %s", vmName, err)
+	}
+
+	if iv.Statuses == nil {
+		return instance.StatusUnknown, nil
+	}
+
+	// it seems that this field should contain two entries: a provisioning status and a power status
+	if len(*iv.Statuses) < 2 {
```

Comment:
So, why not checking for `if len(*iv.Statuses) = 2 {`

Reply:
Because sometimes it doesn't. :P